### ISSUE TITLE
vs2017 packer should set the “Enable NTFS long paths policy”

### ISSUE THREAD

**Issue by perbergland** [Microsoft/azure-pipelines-image-generation#157]
I noticed when putting the agent in a fairly long path (c:\users\admin\downloads\vsts-agent-win-x64-2.134.2\...) that my build failed because of a long path produced in the _work folder.
I then moved it to another folder but also saw that the [NTFS Long Paths policy](https://www.ghacks.net/2016/05/27/microsoft-260-long-path-limit/) was not enabled. Not sure that would have helped, but for the future it would be good if the packer script would run a command to enable that flag so that we can break free from the terrible MAX_PATH problems of the 1990s.

**Comment by perbergland** [comment]
The problem occured when running MSBuild (excerpt below).

I was under the impression that proper long paths was opt in and required both the policy/registry key and an application manifest but I could be wrong.

```
2018-06-04T21:07:15.1579943Z "C:\Users\BokioAdmin\Downloads\vsts-agent-win-x64-2.134.2\_work\1\s\Bokio\InkassogramScheduleCompanyUpdateJob\InkassogramScheduleCompanyUpdateJob.csproj" (default target) (18:10) ->
2018-06-04T21:07:15.1580279Z (CopyAllFilesToSingleFolderForMsdeploy target) -> 
2018-06-04T21:07:15.1580975Z   C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\Web\Microsoft.Web.Publishing.targets(2603,5): error : Copying file C:\Users\BokioAdmin\Downloads\vsts-agent-win-x64-2.134.2\_work\1\s\Bokio\InkassogramScheduleCompanyUpdateJob\bin\Release\Microsoft.IdentityModel.Clients.ActiveDirectory.Platform.dll to obj\Release\Package\PackageTmp\app_data\jobs\triggered\InkassogramScheduleCompanyUpdateJob\Microsoft.IdentityModel.Clients.ActiveDirectory.Platform.dll failed. The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters. [C:\Users\BokioAdmin\Downloads\vsts-agent-win-x64-2.134.2\_work\1\s\Bokio\InkassogramScheduleCompanyUpdateJob\InkassogramScheduleCompanyUpdateJob.csproj]

```

**Comment by perbergland** [comment]
The RT_MANIFEST(1) resource in msbuild.exe does not have the opt-in flag anyway so until msbuild opts in this wouldn't have any effect but still would be a good thing to have in the prepared image already IMO.
Attaching the msbuild manifest and the powershell one for reference.

[msbuild manifest.txt](https://github.com/Microsoft/vsts-image-generation/files/2071091/msbuild.manifest.txt)
[pwsh manifest.txt](https://github.com/Microsoft/vsts-image-generation/files/2071092/pwsh.manifest.txt)

**Comment by vtbassmatt** [comment]
You should be able to [enable this with a regkey](https://betanews.com/2016/05/29/long-paths-windows-10/) to get unblocked.

**Comment by vtbassmatt** [comment]
@bryanmacfarlane do you have any objections to enabling this setting? My understanding is that no tool will be affected unless it explicitly opts in. It won't fix the msbuild issue but I've heard this request from .NET among others, for whom it will fix.

**Comment by vtbassmatt** [comment]
This is about the base Windows image, not about command line tools installed on it. Repathing.

**Comment by DavidStaheli** [comment]
Additional info from [here](https://docs.microsoft.com/en-us/windows/desktop/fileio/naming-a-file#paths):

A registry key allows you to enable or disable the new long path behavior. To enable long path behavior set the registry key at `HKLM\SYSTEM\CurrentControlSet\Control\FileSystem LongPathsEnabled` (Type: `REG_DWORD`). The key's value will be cached by the system (per process) after the first call to an affected Win32 file or directory function (list follows). **The registry key will not be reloaded during the lifetime of the process.** In order for all apps on the system to recognize the value of the key, a reboot might be required because some processes may have started before the key was set.

The registry key can also be controlled via Group Policy at Computer Configuration > Administrative Templates > System > Filesystem > Enable NTFS long paths.

You can also enable the new long path behavior per app via the manifest:

```xml
<application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings xmlns:ws2="https://schemas.microsoft.com/SMI/2016/WindowsSettings">
        <ws2:longPathAware>true</ws2:longPathAware>
    </windowsSettings>
</application>
```
